pkg/kubernetes/scripts/virtual-machines/reducers.jsx

```diff
@@ -36,9 +36,31 @@ const settingsReducer = createReducer([], {
     [actionTypes.SET_SETTINGS]: (state = [], { payload }) => payload ? payload : {}
 })
 
+const vmsMessagesReducer = createReducer({}, {
+    [actionTypes.VM_ACTION_FAILED]: (state = {}, { payload: { vm, message, detail } }) => {
+      const newState = Object.assign({}, state);
+      newState[vm.metadata.uid] = { // So far the last message is kept only
+        message, // textual information
+        detail, // i.e. exception
+      };
+      return newState;
+    },
+
+    [actionTypes.REMOVE_VM_MESSAGE]: (state = {}, { payload: { vm } }) => {
+      if (!state[vm.metadata.uid]) {
+        return state;
+      }
+
+      const newState = Object.assign({}, state);
```

Comment:
What about use of immutable.js ?

Reply:
We can use that.  Let's just make sure we use always the same type (pure js / immutable wrapper) for each property of state tree. I'd like to avoid checks like

```
const baz = foo.bar || for.get('bar')
```

Reply:
Right. Once immutable, always immutable - we would need to avoid mixing it.
The use of immutable is great simplification for reducers.
But its not handy on consumer side (means React component).

So: if reducers are not expected to be complex or their logic has repeatable patterns so it can be easily wrapped to our "home-made" helpers, the immutable.js will just increase complexity and introduce dummy bugs like the one you mentioned.

When thinking about it once more, I think we can try to continue without immutable.js for now. If the complexity grows, we can refactor (no matter it will be ugly job to pass-through whole code)